arango/database.py

```diff
@@ -1217,18 +1218,49 @@ def create_graph(
             parameter cannot be larger than that of **replication_factor**.
             Default value is 1. Used for clusters only.
         :type write_concern: int
+        :param collections: A list collection data objects to provision
+            the graph with. See below for example.
+        :type collections: dict | None
         :return: Graph API wrapper.
         :rtype: arango.graph.Graph
         :raise arango.exceptions.GraphCreateError: If create fails.
 
         Here is an example entry for parameter **edge_definitions**:
 
+        .. code-block:: python
+
+            [
+                {
+                    'edge_collection': 'teach',
+                    'from_vertex_collections': ['teachers'],
+                    'to_vertex_collections': ['lectures']
+                }
+            ]
+
+        Here is an example entry for parameter **collections**:
+        TODO: Rework **collections** data structure?
         .. code-block:: python
 
             {
-                'edge_collection': 'teach',
-                'from_vertex_collections': ['teachers'],
-                'to_vertex_collections': ['lectures']
+                'teachers': {
+                    'docs': teacher_vertices_to_insert
+                    'options': {
+                        'overwrite' = True,
+                        'sync' = True,
+                        'batch_size' = 50
+                    }
+                },
+                'lectures': {
+                    'docs': lecture_vertices_to_insert
+                    'options': {
+                        'overwrite' = False,
+                        'sync' = False,
+                        'batch_size' = 4
+                    }
+                },
+                'teach': {
+                    'docs': teach_edges_to_insert
+                }
```

Comment:
On a closer look, I am not sure if the `collections` parameter makes things easier than just calling the `impulk_bulk` methods on vertex and edge collections separately. In fact, I think it's less convenient because you won't be able to take advantage of argument docstring or type hints from IDEs while composing the dict argument.

Reply:
cc @rajivsam

Reply:
The import bulk methods do not generate the id's for you I believe, so as long as we have a way to automatically create the ID's, that works for me.

Reply:
Is there a strong reason why ID generation cannot live outside of python-arango (e.g. in your application code or in some shared utils package)?

Bulk generating keys from python-arango's side (without proper support from ArangoDB server for atomic operations) doesn't seem like a good idea due to key collision on concurrent requests or retry after failure, unless it's something like uuid4.

Supporting incremental and deterministic key generation for example, means python-arango would have to keep track of ID(s) internally. I think python-arango should aim to be as stateless as possible.

One solution here would be to allow users to pass in a `KeyGenerator` object on collection creation with method like `MyCustomKeyGenerator.generate_next_key()`. This way we would be placing the burden of about managing state and concurrency on the user.

Reply:
I like the key generator idea, works for me!